File: app/src/main/java/ch/bailu/aat/preferences/presets/SolidMET.java
package ch.bailu.aat.preferences.presets;

import android.content.Context;

import java.util.ArrayList;
import java.util.Collections;

import ch.bailu.aat.R;
import ch.bailu.aat.preferences.OldSolidMET;
import ch.bailu.aat.preferences.SolidString;
import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat.exception.ValidationException;

public class SolidMET extends SolidString {
    private final int preset;

    public SolidMET(Context c, int i) {
        super(c, SolidMET.class.getSimpleName()+ "_" + i);
        preset = i;
    }

    @Override
    public String getLabel() {
        return getContext().getString(R.string.p_met);
    }

    @Override
    public String getValueAsString() {
        String r = super.getValueAsString();

        if (Storage.DEF_VALUE.equals(r)) {
            r = new OldSolidMET(getContext(), preset).getValueAsString();
        }
        return r;
    }


    public float getMETValue() {
        String val = getValueAsString();

        final int from = 0;
        int to = val.indexOf(' ');

        float r = 0f;


        if (to > from) {
            try {
                String met = val.substring(from, to);
                r = Float.valueOf(met);
            } catch (NumberFormatException e) {
                // TODO : userfeedback missing
                r = 0f;
            }
        }

        if (r > 20f || r < 0f) {
            // TODO : userfeedback missing, validation should be moved to MET preferences
            r = 0f;
        }
        return r;
    }

    @Override
    public void setValueFromString(String s) throws ValidationException {
        s = s.trim();

        if (! validate(s)) {
            throw new ValidationException(getString(R.string.error_met));
        } else {
            setValue(s);
        }
    }


    @Override
    public ArrayList<String> buildSelection(ArrayList<String> list) {

        String[] array = getContext().getResources().getStringArray(R.array.p_met_list);

        Collections.addAll(list, array);

        return list;
    }

    @Override
    public boolean validate(String s) {
        // entering 0.0 is still possible
        return s.split(" ")[0].matches("1?[0-9].[0-9]|20.0");
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/preferences/presets/SolidMET.kt
package ch.bailu.aat_lib.preferences.presets

import ch.bailu.aat_lib.exception.ValidationException
import ch.bailu.aat_lib.preferences.SolidString
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.resources.Res
import java.util.Collections
import javax.annotation.Nonnull

class SolidMET(storage: StorageInterface, private val preset: Int) : SolidString(
    storage, SolidMET::class.java.simpleName + "_" + preset
) {
    @Nonnull
    override fun getLabel(): String {
        return Res.str().p_met()
    }

    @Nonnull
    override fun getValueAsString(): String {
        var result = super.getValueAsString()
        if (getStorage().isDefaultString(result)) {
            result = defaultValue
        }
        return result
    }

    private val defaultValue: String
        get() {
            val metList = Res.str().p_met_list()
            return if (preset < metList.size) {
                metList[preset]
            } else {
                metList[0]
            }
        }

    val metValue: Float
        get() {
            val `val` = getValueAsString()
            val from = 0
            val to = `val`.indexOf(' ')
            var r = 0f
            if (to > from) {
                r = try {
                    val met = `val`.substring(from, to)
                    met.toFloat()
                } catch (e: NumberFormatException) {
                    0f
                }
            }
            if (r > 20f || r < 0f) {
                r = 0f
            }
            return r
        }

    @Throws(ValidationException::class)
    override fun setValueFromString(string: String) {
        val trimmedString = string.trim { it <= ' ' }
        if (validate(trimmedString)) {
            setValue(trimmedString)
        } else {
            throw ValidationException(Res.str().error_met())
        }
    }

    override fun buildSelection(list: ArrayList<String>): ArrayList<String> {
        val array = Res.str().p_met_list()
        Collections.addAll(list, *array)
        return list
    }

    override fun validate(s: String): Boolean {
        // entering 0.0 is still possible
        return s.split(" ".toRegex()).dropLastWhile { it.isEmpty() }
            .toTypedArray()[0].matches("1?[0-9].[0-9]|20.0".toRegex())
    }
}
